app/models/test_track/session.rb

```diff
@@ -11,34 +11,25 @@ def initialize(controller)
   def manage
     yield
   ensure
-    manage_cookies!
-    manage_response_headers!
+    if visitor.id_loaded?
+      manage_cookies!
+      manage_response_headers!
+    end
     notify_unsynced_assignments! if sync_assignments?
   end
 
   def visitor_dsl_for(identity)
-    if has_matching_identity?(identity)
-      visitor_dsl
-    else
-      TestTrack::VisitorDSL.new(visitors_by_identity[identity])
-    end
+    TestTrack::VisitorDSL.new(visitors_by_identity[identity])
   end
 
   def visitors_by_identity
     @visitors_by_identity ||= Hash.new do |visitors_by_identity, identity|
-      remote_visitor = TestTrack::Remote::Visitor.from_identifier(
-        identity.test_track_identifier_type,
-        identity.test_track_identifier_value
-      )
-      visitors_by_identity[identity] = TestTrack::Visitor.new(
-        id: remote_visitor.id,
-        assignments: remote_visitor.assignments
-      )
+      visitors_by_identity[identity] = TestTrack::LazyVisitorByIdentity.new(identity)
```

Comment:
This tidies up by delegating to the lazy visitor.

Reply:
I still think it would be nice to have an object instead of relying on the Hash default proc thing here.

Reply:
Ok, I’ll try that. The lazy visitor has enough responsibilities though and I thought this got skinny enough to be idiomatic as a hash default, but it’s still pretty heady.